presto-pinot/src/main/java/io/prestosql/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,433 @@
+    private final List<PinotColumnHandle> columnHandles;
+    private final PinotClusterInfoFetcher clusterInfoFetcher;
+    private final ConnectorSession session;
+    private final ObjectMapper objectMapper;
+
+    private boolean finished;
+    private long readTimeNanos;
+    private long completedBytes;
+
+    public PinotBrokerPageSource(
+            ConnectorSession session,
+            BrokerPqlWithContext brokerPql,
+            List<PinotColumnHandle> columnHandles,
+            PinotClusterInfoFetcher clusterInfoFetcher,
+            ObjectMapper objectMapper)
+    {
+        this.brokerPql = requireNonNull(brokerPql, "broker is null");
+        this.clusterInfoFetcher = requireNonNull(clusterInfoFetcher, "cluster info fetcher is null");
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.session = requireNonNull(session, "session is null");
+        this.objectMapper = requireNonNull(objectMapper, "object mapper is null");
+    }
+
+    private static Double parseDouble(String value)
+    {
+        try {
+            return Double.valueOf(value);
+        }
+        catch (NumberFormatException ne) {
+            switch (value) {
+                case PINOT_INFINITY:
+                case PINOT_POSITIVE_INFINITY:
+                    return PRESTO_INFINITY;
+                case PINOT_NEGATIVE_INFINITY:
+                    return PRESTO_NEGATIVE_INFINITY;
+            }
+            throw new PinotException(PINOT_DECODE_ERROR, Optional.empty(), "Cannot decode double value from pinot " + value, ne);
+        }
+    }
+
+    private void setValue(Type type, BlockBuilder blockBuilder, String value)
+    {
+        if (value == null) {
+            blockBuilder.appendNull();
+            return;
+        }
+        if (!(type instanceof FixedWidthType) && !(type instanceof VarcharType)) {
+            throw new PinotException(PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "type '" + type + "' not supported");
+        }
+        if (type instanceof FixedWidthType) {
+            completedBytes += ((FixedWidthType) type).getFixedSize();
+            if (type instanceof BigintType) {
+                type.writeLong(blockBuilder, parseDouble(value).longValue());
+            }
+            else if (type instanceof IntegerType) {
+                blockBuilder.writeInt(parseDouble(value).intValue());
+            }
+            else if (type instanceof TinyintType) {
+                blockBuilder.writeByte(parseDouble(value).byteValue());
+            }
+            else if (type instanceof SmallintType) {
+                blockBuilder.writeShort(parseDouble(value).shortValue());
+            }
+            else if (type instanceof BooleanType) {
+                type.writeBoolean(blockBuilder, parseBoolean(value));
+            }
+            else if (type instanceof DecimalType || type instanceof DoubleType) {
+                type.writeDouble(blockBuilder, parseDouble(value));
```

Comment:
This is incorrect for DecimalType. You need to do something like this for a "long" decimal (i.e., > 64 bits):

```java
type.writeSlice(blockBuilder, doubleToLongDecimal(parseDouble(value), type.getPrecision(), type.getScale()));
```

and, for a "short" decimal (<= 64 bits):

```java
type.writeLong(blockBuilder, doubleToShortDecimal(parseDouble(value), type.getPrecision(), type.getScale()));
```

See `io.prestosql.spi.type.DecimalConversions`

Reply:
Removed, pinot doesn't have a decimal type